File: downloads/code/LeetCode/Java/Combinations.java
public class Solution {
    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> ret = new LinkedList<List<Integer>> ();
        List<Integer> tmp = new LinkedList<Integer>();

        dfs(ret, tmp, 1, k, n); 
        return ret;       
    }

    public void dfs(List<List<Integer>> ret, List<Integer> tmp, int start, int k, int n){
        if(tmp.size() == k){
            ret.add(new LinkedList(tmp));
            return;
        }
        for(int i=start;i<=n;i++){
            tmp.add(i);
            dfs(ret, tmp, i+1,k,n);
            tmp.remove(tmp.size()-1);
        }
    }
}

File: downloads/code/LeetCode/Python/Combinations.py
class Solution(object):
    def recur_combine(self, result, vector, i, n, k):
        if len(vector) == k:
            result.append(vector[:])
            return
        
        while i <= n:
            vector.append(i)
            self.recur_combine(result,vector,i+1,n,k)
            vector.pop()
            i = i+1
        return 


    def combine(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[List[int]]
        """
        
        result = []
        if n < k:
            return result
        self.recur_combine(result, [], 1, n, k)
        return result
